File: scripts/verify-single-variable-calculus.mjs
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import assert from 'node:assert/strict';
import { spawnSync } from 'node:child_process';
import { parse } from '@babel/parser';
import katex from 'katex';
import * as model from '../src/learn/data/single-variable-calculus-models.js';
import { singleVariableCalculusExamples } from '../src/learn/data/single-variable-calculus-examples.js';

const directory = 'scratch/single-variable-calculus-verification';
fs.mkdirSync(directory, { recursive: true });
const data = { rates: [], limits: [], compositions: [], extrema: [], accumulations: [], growth: [], taylors: [], improper: [], examples: singleVariableCalculusExamples };
for (let t = 4; t <= 12; t++) for (const h of [-1, -.5, -.1, -.01, -.0001, .0001, .01, .1, .5, 1]) data.rates.push(model.motionRate(t / 4, h));
for (const kind of ['smooth', 'hole', 'jump']) for (const epsilon of [.01, .03, .1, .3, .5, .83, .84, .8399999999999999, .8400000000000001, 1]) for (const delta of [.001, .002, .01, .02, .1, .2, .25, .5]) data.limits.push(model.limitGuarantee(kind, epsilon, delta));
for (const x of [-1, -.5, -1 / 3, 0, .25, .5, 1]) for (const h of [-.5, -.1, -.01, 0, .01, .1, .5]) data.compositions.push(model.compositionChange(x, h));
for (const kind of ['motion', 'inflection', 'cusp']) for (let a = 0; a < 16; a++) for (let b = a + 1; b <= 16; b++) data.extrema.push(model.extremaCandidates(kind, a / 4, b / 4));
for (const kind of ['motion', 'inflection', 'cusp']) for (const [a,b] of [[2,2+2*Number.EPSILON],[1,1+Number.EPSILON],[0,1+Number.EPSILON],[0,1e-323],[3-2*Number.EPSILON,3]]) data.extrema.push(model.extremaCandidates(kind,a,b));
for (let upper = 1; upper <= 16; upper++) for (const n of [1, 2, 3, 4, 8, 16, 32, 64, 127, 128]) for (const tag of ['left', 'midpoint', 'right']) data.accumulations.push(model.motionAccumulation(upper / 4, n, tag));
for (let rate = -16; rate <= 16; rate++) for (let period = 1; period <= 8; period++) data.growth.push(model.exponentialRate(rate / 20, period / 4));
for (const kind of ['exp', 'log']) for (let degree = 0; degree <= 12; degree++) for (const x of (kind === 'exp' ? [-2, -1.9, -1, -.25, 0, .01, .5, 1, 2] : [-.9, -.8, -.1, 0, .01, .5, .9, 1, 1.05, 1.5])) data.taylors.push(model.taylorApproximation(kind, degree, x));
for (const kind of ['tail', 'endpoint']) for (const power of [0, .5, .9, 1 - Number.EPSILON, 1, 1 + Number.EPSILON, 1.5, 2, 3]) for (let q = 0; q <= 24; q++) data.improper.push(model.improperPowerIntegral(kind, power, q / 4));
for(const x of [.05,.1,.15,.2]) data.taylors.push(model.taylorApproximation('exp',12,x));
const invalid = [
  () => model.motionPosition(-1), () => model.motionVelocity(Infinity), () => model.motionRate(0, .1), () => model.motionRate(2, 0), () => model.motionRate(2, 1e-10),
  () => model.limitGuarantee('unknown', .1, .1), () => model.limitGuarantee('smooth', NaN, .1), () => model.limitGuarantee('hole', .1, 0),
  () => model.compositionChange(2, .1), () => model.compositionChange(0, Infinity), () => model.extremaCandidates('motion', 2, 2), () => model.extremaCandidates('cusp', 4, 1), () => model.extremaCandidates('bad', 0, 4),
  () => model.motionAccumulation(4, 129, 'left'), () => model.motionAccumulation(4, 1.5, 'left'), () => model.motionAccumulation(.1, 4, 'left'), () => model.motionAccumulation(4, 4, 'bad'),
  () => model.exponentialRate(.9, 1), () => model.exponentialRate(0, 0), () => model.taylorApproximation('log', 2, -1), () => model.taylorApproximation('exp', 13, 0), () => model.taylorApproximation('bad', 2, 0),
  () => model.improperPowerIntegral('bad', 1, 1), () => model.improperPowerIntegral('tail', -1, 1), () => model.improperPowerIntegral('tail', 1, 7),
];
invalid.forEach(call => assert.throws(call, RangeError));
const source = 'src/learn/data/topics/single-variable-calculus-limits-derivatives-integrals.jsx';
const ast = parse(fs.readFileSync(source, 'utf8'), { sourceType: 'module', plugins: ['jsx'] });
const formulas = [];
function visit(node) {
  if (!node || typeof node !== 'object') return;
  if (node.type === 'JSXElement' && node.openingElement.name?.name === 'MathBlock') {
    const expression = node.children.find(item => item.type === 'JSXExpressionContainer').expression;
    assert.equal(expression.type, 'TaggedTemplateExpression');
    assert.equal(expression.quasi.expressions.length, 0);
    const tex = expression.quasi.quasis.map(item => item.value.raw).join('');
    katex.renderToString(tex, { throwOnError: true, displayMode: true });
    formulas.push(tex);
  }
  for (const [key, value] of Object.entries(node)) if (!['loc', 'start', 'end'].includes(key)) {
    if (Array.isArray(value)) value.forEach(visit); else if (value && typeof value === 'object') visit(value);
  }
}
visit(ast);
assert(formulas.length > 20);
data.syntax = { equations: formulas.length, invalidRejected: invalid.length };
data.sources = [source, 'src/learn/data/single-variable-calculus-models.js', 'src/learn/data/single-variable-calculus-examples.js', 'src/learn/components/lesson-labs/SingleVariableCalculusLabs.jsx', 'src/learn/components/lesson-labs/single-variable-calculus-labs.css', 'src/learn/data/curriculum/blueprints/single-variable-calculus-limits-derivatives-integrals.js'].map(file => ({ path: file, sha256: crypto.createHash('sha256').update(fs.readFileSync(file)).digest('hex') }));
fs.writeFileSync(`${directory}/cases.json`, JSON.stringify(data));
const result = spawnSync(path.resolve('scratch/lesson-tools/Scripts/python.exe'), ['scripts/verify-single-variable-calculus.py'], { encoding: 'utf8' });
process.stdout.write(result.stdout); process.stderr.write(result.stderr);
if (result.status !== 0) process.exit(result.status || 1);

File: scripts/verify-single-variable-calculus.py
"""Exact and high precision oracles for the actual bounded models and programs."""
import contextlib
import io
import json
import math
import platform
from datetime import datetime, timezone
from fractions import Fraction as F
from pathlib import Path

import mpmath as mp
import sympy as sp

mp.mp.dps = 65
directory = Path('scratch/single-variable-calculus-verification')
data = json.loads((directory / 'cases.json').read_text(encoding='utf-8'))


def close(actual, expected, tolerance=3e-12):
    assert math.isclose(actual, float(expected), rel_tol=tolerance, abs_tol=tolerance), (actual, expected)


def fraction(value):
    return F(str(value))


def position(t):
    return t**3 - 6*t**2 + 9*t


for s in data['rates']:
    t, h = fraction(s['baseTime']), fraction(s['increment'])
    change = position(t+h)-position(t)
    derivative = 3*t*t-12*t+9
    for field, expected in [('position', position(t)), ('nextPosition', position(t+h)), ('velocity', derivative), ('secant', change/h), ('acceleration', 6*t-12), ('predictedChange', derivative*h), ('exactChange', change), ('remainder', change-derivative*h)]:
        close(s[field], expected)

for s in data['limits']:
    epsilon, delta = fraction(s['epsilon']), fraction(s['delta'])
    jump = 2 if s['kind'] == 'jump' else 0
    supremum = (2+delta)**2-4+jump
    assert s['guaranteed'] == (supremum <= epsilon)
    close(s['supremum'], supremum)
    if s['guaranteed']:
        assert s['witness'] is None
    else:
        w = s['witness']
        x = F(w['inputExact'])
        assert 0 < x-2 < delta
        assert F(w['displacementExact']) == x-2
        assert F(w['errorExact']) == x*x-4+jump >= epsilon
        if w['resolvedInFloat']:
            assert 0 < abs(w['input']-2) < float(delta) and w['error'] >= float(epsilon)

for s in data['compositions']:
    x, h = fraction(s['input']), fraction(s['increment'])
    f = lambda z: 9*z*z+6*z+1
    derivative = 18*x+6
    for field, expected in [('output', f(x)), ('nextOutput', f(x+h)), ('intermediate', 3*x+1), ('intermediateChange', 3*h), ('outerDerivative', 6*x+2), ('derivative', derivative), ('predictedChange', derivative*h), ('exactChange', f(x+h)-f(x)), ('remainder', 9*h*h)]:
        close(s[field], expected)

t = sp.Symbol('t', real=True)
polynomials = {'motion': t**3-6*t**2+9*t, 'inflection': (t-2)**3}
for s in data['extrema']:
    a, b = fraction(s['left']), fraction(s['right'])
    if s['kind'] == 'cusp':
        critical = [F(2)]
        f = lambda x: abs(x-2)
        derivative = lambda x: None if x == 2 else (-1 if x < 2 else 1)
    else:
        expression = polynomials[s['kind']]
        differentiated = sp.diff(expression, t)
        critical = [F(str(root)) for root in sp.solve(differentiated, t)]
        f = lambda x: F(expression.subs(t, sp.Rational(x.numerator, x.denominator)))
        derivative = lambda x: F(differentiated.subs(t, sp.Rational(x.numerator, x.denominator)))
    candidates = sorted(set([a, b]+[x for x in critical if a <= x <= b]))
    assert [fraction(c['input']) for c in s['candidates']] == candidates
    values = [f(x) for x in candidates]
    close(s['minimum'], min(values)); close(s['maximum'], max(values))
    for c, x, value in zip(s['candidates'], candidates, values):
        close(c['value'], value)
        assert F(c['valueExact']) == value
        assert c['isMinimum'] == (value == min(values))
        assert c['isMaximum'] == (value == max(values))
        if derivative(x) is None: assert c['derivative'] is None
        else: close(c['derivative'], derivative(x))
    for interval in s['intervals']:
        midpoint = (fraction(interval['start'])+fraction(interval['end']))/2
        d = derivative(midpoint)
        assert interval['sign'] == (1 if d > 0 else -1 if d < 0 else 0)

for s in data['accumulations']:
    upper, n = fraction(s['upper']), s['panelCount']
    width = upper/n
    offset = {'left': F(0), 'midpoint': F(1,2), 'right': F(1)}[s['method']]
    terms = []
    for index, panel in enumerate(s['panels']):
        x = (index+offset)*width
        height = 3*x*x-12*x+9
        terms.append(width*height)
        for field, expected in [('start', index*width), ('end', (index+1)*width), ('sample', x), ('height', height), ('signedContribution', width*height)]:
            close(panel[field], expected)
    close(s['signedEstimate'], sum(terms)); close(s['distanceEstimate'], sum(map(abs, terms)))
    close(s['exactDisplacement'], position(upper))
    # Integrate positive/negative polynomial branches symbolically, splitting at roots.
    cuts = [F(0)]+[F(x) for x in (1,3) if x < upper]+[upper]
    distance = sum(abs(position(b)-position(a)) for a,b in zip(cuts,cuts[1:]))
    close(s['exactDistance'], distance)
    close(s['signedError'], sum(terms)-position(upper))
    close(s['endpointVelocity'], 3*upper**2-12*upper+9)

for s in data['growth']:
    k, period = mp.mpf(str(s['rate'])), mp.mpf(str(s['period']))
    current, future = 10*mp.exp(2*k), 10*mp.exp((2+period)*k)
    for field, expected in [('currentAmount', current), ('futureAmount', future), ('instantaneousRate', k*current), ('intervalAverageRate', (future-current)/period), ('perPeriodFraction', mp.exp(k*period)-1), ('averageRelativeRate', (mp.exp(k*period)-1)/period), ('tangentPrediction', current+k*current*period)]:
        close(s[field], expected)

for s in data['taylors']:
    x, degree = mp.mpf(str(s['input'])), s['degree']
    f = mp.exp if s['kind'] == 'exp' else lambda z: mp.log(1+z)
    coefficients = mp.taylor(f, 0, degree)
    polynomial = mp.polyval(list(reversed(coefficients)), x)
    close(s['polynomial'], polynomial); close(s['exactValue'], f(x))
    close(s['signedError'], polynomial-f(x))
    maximum = mp.exp(max(0,x)) if s['kind'] == 'exp' else mp.factorial(degree)/min(1,1+x)**(degree+1)
    bound = maximum*abs(x)**(degree+1)/mp.factorial(degree+1)
    close(s['remainderBound'], bound)
    assert abs(polynomial-f(x)) <= bound*(1+mp.mpf('1e-55'))
    if degree in (0,2,6) and x in (mp.mpf('-.9'),mp.mpf('.5'),mp.mpf('1.5')):
        remainder = mp.quad(lambda u: mp.diff(f,u,degree+1)*(x-u)**degree/mp.factorial(degree), [0,x])
        assert abs(remainder-(f(x)-polynomial)) < mp.mpf('1e-48')

for s in data['improper']:
    # Near p=1, preserve the binary float input exactly. Its distance to1 matters.
    power, q = mp.mpf(s['power']), mp.mpf(s['cutoffExponent'])
    log_cutoff = q*mp.log(10)
    sign = 1 if s['kind'] == 'tail' else -1
    cutoff = mp.exp(sign*log_cutoff)
    # x=exp(u) converts the definite integral to an exponential integrand.
    a,b = (0,log_cutoff) if sign == 1 else (-log_cutoff,0)
    truncated = mp.quad(lambda u: mp.exp((1-power)*u), [a,b]) if a != b else 0
    close(s['cutoff'], cutoff); close(s['truncated'], truncated)
    converges = power > 1 if sign == 1 else power < 1
    assert s['converges'] == converges
    if converges:
        total = 1/abs(power-1)
        close(s['total'], total); close(s['missing'], total-truncated)
    else: assert s['total'] is None and s['missing'] is None

programs, namespaces = [], {}
for example in data['examples']:
    capture, namespace = io.StringIO(), {'__name__': '__main__'}
    with contextlib.redirect_stdout(capture):
        exec(compile(example['code'], example['id']+'.py', 'exec'), namespace)
    assert capture.getvalue().strip() == example['expected'].strip(), example['id']
    programs.append(example['id']); namespaces[example['id']] = namespace
for x in [F(-3,2),F(1,3),F(5,2),F(7)]:
    assert namespaces['motion-secants']['position'](x) == position(x)
    assert namespaces['interval-extrema']['position'](x) == position(x)
    assert namespaces['signed-rectangles']['velocity'](x) == 3*x*x-12*x+9
    assert namespaces['moving-bounds']['accumulated'](x) == (x*x+(x*x)**3/F(3))-(x+x**3/F(3))

# Changed practice and displayed application values, independent symbolic operations.
x = sp.Symbol('x', real=True)
assert sp.diff((2*x-1)**2,x).subs(x,0) == -4
assert sp.Rational(4,5)**2-1 == -sp.Rational(9,25)
assert sp.integrate(2*x-4,(x,0,4)) == 0
assert -sp.integrate(2*x-4,(x,0,2))+sp.integrate(2*x-4,(x,2,4)) == 8
assert sp.simplify(sp.integrate(2*x/(1+x*x),(x,1,3))-sp.log(5)) == 0
assert sp.integrate(x*sp.sin(x),(x,0,sp.pi)) == sp.pi
assert sp.integrate(3+2*x,(x,0,2)) == 10
assert sp.integrate(2+x,(x,0,2)) == 6
assert sp.integrate(x*(2+x),(x,0,2)) == sp.Rational(20,3)
assert sp.integrate(sp.pi*x,(x,0,3)) == 9*sp.pi/2
assert sp.integrate(sp.sqrt(1+x),(x,0,3)) == sp.Rational(14,3)
assert sp.integrate(x*sp.cos(x),(x,0,1)) == sp.sin(1)+sp.cos(1)-1
close(math.log(.8)/3,-.0743812,1e-7)
close(math.log(.5)/(math.log(.8)/3),9.318851,1e-7)
assert 0 < mp.mpf('3.05')-mp.sqrt(mp.mpf('9.3')) < mp.mpf(1)/2400
close(mp.mpf(2)/mp.sqrt(100), .2)
result = {'at': datetime.now(timezone.utc).isoformat(), 'passed': True, 'python': platform.python_version(), 'mpmathPrecision': mp.mp.dps,
          'cases': {key:len(value) for key,value in data.items() if isinstance(value,list) and key not in ('sources','examples')}, 'actualPrograms': programs, 'changedNativeHelpers': 16, 'practiceAndApplicationChecks': 16, 'syntax':data['syntax'], 'sources':data['sources'],
          'limits':['Finite and symbolic checks support the implementations; they do not replace the written real-analysis proofs.', 'Browser geometry, readable rendering and production loading have separate records.']}
(directory/'results.json').write_text(json.dumps(result,indent=2)+'\n',encoding='utf-8')
print(json.dumps(result))
